File: BreadLang/Parser/Nodes/SubRoutineNode.cs
﻿using BreadLang.Compiling;
using BreadLang.Parser.Nodes.InstructionNodes;
using BreadLang.Tokens;

namespace BreadLang.Parser.Nodes;

public class SubRoutineNode(ProgramNode programNode, string name) : Node
{
	public string Name { get; private set; } = name;
	public override void Populate(Parser parser)
	{
		parser.SkipNewLines();

		// expect indent token
		var indent = parser.Expect(TokenType.Indent);

		if (!programNode.CheckIndent(indent))
		{
			ErrorHandler.Instance!.Error(indent, "Invalid indent");
		}

		while (!parser.IsAtEnd())
		{
			var token = parser.Advance();
			Node? node = null;
			switch (token.Type)
			{
				case TokenType.NewLine:
					parser.SkipNewLines();

					if (!programNode.CheckIndent(parser.Peek()))
						return;
					
					parser.Advance();
					break;

				// instructions
				case TokenType.Nop: node = new NopNode(); break;
				case TokenType.Lw: node = new LwNode(); break;
				case TokenType.Sw: node = new SwNode(); break;
				case TokenType.Mw: node = new MwNode(); break;
				case TokenType.Push: node = new PushNode(); break;
				case TokenType.Pop: node = new PopNode(); break;
				case TokenType.Lda: node = new LdaNode(); break;
				case TokenType.Jmp: node = new JmpNode(); break;
				case TokenType.Jz: node = new JzNode(); break;
				case TokenType.Jo: node = new JoNode(); break;
				case TokenType.Add: node = new AddNode(); break;
				case TokenType.Sub: node = new SubNode(); break;
				case TokenType.Out: node = new OutNode(); break;
				case TokenType.Hlt: node = new HltNode(); break;

				case TokenType.Eof:
					return;

				case TokenType.Identifier:
					// check if its a macro
					if (!programNode.Macros.ContainsKey(token.Lexeme))
					{
						ErrorHandler.Instance!.Error(token, $"Unexpected token {token.Type}. Expected an instruction or macro");
						break;
					}

					// capture the arguments of the macro
					List<Token> arguments = new();

					while (true)
					{
						arguments.Add(parser.Advance());

						// capture the comma
						if (!parser.Check(TokenType.Comma))
						{
							break;
						}
						parser.Advance(); // capture the comma
					}

					if (!parser.Check(TokenType.NewLine) && !parser.Check(TokenType.Eof))
						ErrorHandler.Instance!.Error(token, $"Incorrect macro usage. Expected new line or end of file");

					var macro = programNode.Macros[token.Lexeme];
					var macroTokens = macro.GetTokens(arguments);
					parser.InsertTokens(macroTokens);
					break;

				default:
					ErrorHandler.Instance!.Error(token, $"Unexpected token {token.Type}. Expected an instruction or macro");
					return;
			}

			if (node == null) continue;

			node.Populate(parser);
			Children.Add(node);
		}
	}

	public override void Compile(Compiler compiler)
	{
		foreach (var child in Children)
		{
			child.Compile(compiler);
		}
	}

	public override string ToString()
	{
		return $"Subroutine({name})";
	}

	public override int GetSize()
	{
		int size = 0;

		foreach (Node node in Children)
		{
			size += node.GetSize();
		}

		return size;
	}
}

File: src/compiler/parser/subroutine_node.rs
use crate::compiler::lexer::TokenType;

use super::{instruction_node::InstructionNode, Node, Parser};

#[derive(Debug)]
pub struct SubroutineNode {
    name: String,
    instructions: Vec<InstructionNode>
}

impl Node for SubroutineNode {
    fn populate(parser: &mut Parser) -> Self {
        parser.skip_new_lines();

        // identifier
        let identifier = parser.advance();
        let name: String;
        if let TokenType::Identifier(n) = &identifier.token_type {
            name = n.clone();
        } else {
            panic!("Expected identifier. Found {:?}", identifier.token_type)
        }
        
        // expect colon
        if !matches!(parser.advance().token_type, TokenType::Colon) {
            panic!("Expected colon {:?}", parser.current().token_type)
        }

        // expect new line
        if !matches!(parser.advance().token_type, TokenType::NewLine) {
            panic!("Expected new line")
        }

        let mut instructions: Vec<InstructionNode> = Vec::new();

        while !parser.is_at_end() {
            parser.skip_new_lines();
            if !matches!(parser.peek().token_type, TokenType::Indent(_)) {
                break;
            }
            
            parser.advance(); // advance past indent

            instructions.push(InstructionNode::populate(parser));
        }

        SubroutineNode {
            name,
            instructions
        }
    }

    fn get_size(&self) -> i32 {
        let mut size = 0;
        
        for node in &self.instructions {
            size += node.get_size();
        }

        size
    }
}
